src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,560 @@
+  // Note: today, tomorrow, yesterday are time zone related, should use time zone to generate.
+  cudf::timestamp_us epoch     = cudf::timestamp_us{cudf::duration_us{111L}};
+  cudf::timestamp_us now       = cudf::timestamp_us{cudf::duration_us{222L}};
+  cudf::timestamp_us today     = cudf::timestamp_us{cudf::duration_us{333L}};
+  cudf::timestamp_us tomorrow  = cudf::timestamp_us{cudf::duration_us{444L}};
+  cudf::timestamp_us yesterday = cudf::timestamp_us{cudf::duration_us{555L}};
+
+  __device__ thrust::tuple<cudf::timestamp_us, bool> operator()(const cudf::size_type& idx) const
+  {
+    auto const d_str = d_strings.element<cudf::string_view>(idx);
+    return parse_string_to_timestamp_us(d_str,
+                                        default_time_zone,
+                                        default_time_zone_char_len,
+                                        allow_time_zone,
+                                        allow_special_expressions,
+                                        epoch,
+                                        now,
+                                        today,
+                                        tomorrow,
+                                        yesterday);
+  }
+};
+
+/**
+ *
+ * Trims and parses timestamp string column to a timestamp column and a is valid column
+ *
+ */
+std::pair<std::unique_ptr<cudf::column>, std::unique_ptr<cudf::column>> to_timestamp(
+  cudf::strings_column_view const& input,
+  std::string_view const& default_time_zone,
+  bool allow_time_zone,
+  bool allow_special_expressions,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  auto d_strings = cudf::column_device_view::create(input.parent(), stream);
+
+  auto output_timestamp =
+    cudf::make_timestamp_column(cudf::data_type{cudf::type_id::TIMESTAMP_MICROSECONDS},
+                                input.size(),
+                                cudf::detail::copy_bitmask(input.parent(), stream, mr),
+                                input.null_count(),
+                                stream,
+                                mr);
+  // record which string is failed to parse.
+  auto output_bool =
+    cudf::make_fixed_width_column(cudf::data_type{cudf::type_id::BOOL8},
+                                  input.size(),
+                                  cudf::detail::copy_bitmask(input.parent(), stream, mr),
+                                  input.null_count(),
+                                  stream,
+                                  mr);
+
+  thrust::transform(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator(0),
+    thrust::make_counting_iterator(input.size()),
+    thrust::make_zip_iterator(
+      thrust::make_tuple(output_timestamp->mutable_view().begin<cudf::timestamp_us>(),
+                         output_bool->mutable_view().begin<bool>())),
+    parse_timestamp_string_fn{*d_strings,
+                              default_time_zone.data(),
```

Comment:
This is a std::string_view, it's in CPU memory, should copy to GPU.
Alfred will fix this in a follow-on commit.